OpenRA.Mods.Common/Graphics/RailgunRenderable.cs

```diff
@@ -0,0 +1,89 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Drawing;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Projectiles;
+
+namespace OpenRA.Mods.Common.Graphics
+{
+	public struct RailgunHelixRenderable : IRenderable, IFinalizedRenderable
+	{
+		readonly WPos pos;
+		readonly int zOffset;
+		readonly Railgun railgun;
+		readonly RailgunInfo info;
+		readonly int helixRadius;
+		readonly int alpha;
+		readonly int ticks;
+
+		WAngle angle;
+
+		// Not-so OOP function signature but this is a game, must save computations!
+		public RailgunHelixRenderable(WPos pos, int zOffset, Railgun railgun, RailgunInfo railgunInfo, int ticks)
+		{
+			this.pos = pos;
+			this.zOffset = zOffset;
+			this.railgun = railgun;
+			this.info = railgunInfo;
+			this.ticks = ticks;
+
+			helixRadius = info.HelixRadius + ticks * info.HelixRadiusDeltaPerTick;
+			alpha = 128 - ticks * 8;
+			if (alpha < 0)
+				alpha = 0;
+			angle = new WAngle(ticks * info.HelixAngleDeltaPerTick);
+		}
+
+		public WPos Pos { get { return pos; } }
+		public PaletteReference Palette { get { return null; } }
+		public int ZOffset { get { return zOffset; } }
+		public bool IsDecoration { get { return true; } }
+
+		public IRenderable WithPalette(PaletteReference newPalette) { return new RailgunHelixRenderable(pos, zOffset, railgun, info, ticks); }
+		public IRenderable WithZOffset(int newOffset) { return new RailgunHelixRenderable(pos, newOffset, railgun, info, ticks); }
+		public IRenderable OffsetBy(WVec vec) { return new RailgunHelixRenderable(pos + vec, zOffset, railgun, info, ticks); }
+		public IRenderable AsDecoration() { return this; }
+
+		public IFinalizedRenderable PrepareRender(WorldRenderer wr) { return this; }
+		public void Render(WorldRenderer wr)
+		{
+			if (railgun.ForwardStep == WVec.Zero)
+				return;
+
+			var screenWidth = wr.ScreenVector(new WVec(info.HelixThichkess, 0, 0))[0];
+
+			// last point the rad beam "reached"
+			var pos = this.pos; // where we are.
+			float3 last = wr.Screen3DPosition(pos);
+			for (var i = railgun.CycleCnt * info.QuantizationCount; i > 0; i--)
+			{
+				// Make it narrower near the end.
+				var rad = i < info.QuantizationCount ? helixRadius / 4 :
+					i < 2 * info.QuantizationCount ? helixRadius / 2 :
+					helixRadius;
+
+				var u = rad * angle.Cos() * railgun.LeftVector / (1024 * 1024)
+					+ rad * angle.Sin() * railgun.UpVector / (1024 * 1024);
+				var end = wr.Screen3DPosition(pos + u);
+
+				Game.Renderer.WorldRgbaColorRenderer.DrawLine(last, end, screenWidth, Color.FromArgb(alpha, railgun.HelixColor));
```

Comment:
I haven't tested this PR, but I expect that rendering individual line segments with alpha will cause artefacts at the joins where they overlap.  Pull this outside the loop and use the `IEnumerable<float3>` overload to have the renderer properly join up the segments.